cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/AbstractStreamWriterSizeManager.java

```
@@ -0,0 +1,147 @@
+                                         new StreamWriterHeartbeat(System.currentTimeMillis(), entry.getValue(),
+                                                                   instanceId, StreamWriterHeartbeat.Type.REGULAR))
+          .get();
+      } catch (Throwable t) {
+        LOG.error("Could not publish final heartbeat for stream {} on writer instance {}",
+                  entry.getKey(), instanceId, t);
+      }
+    }
+
+    heartbeatPublisher.stopAndWait();
+  }
+
+  @Override
+  public void received(String streamName, long dataSize) {
+    if (dataSize <= 0) {
+      return;
+    }
+
+    boolean success;
+    do {
+      Long currentSize = absoluteSizes.get(streamName);
+      if (currentSize == null) {
+        success = absoluteSizes.putIfAbsent(streamName, dataSize) == null;
+        if (success) {
+          // This thread successfully put the name of a new Stream in the map,
+          // it can schedule the heartbeats for this stream too.
+          scheduleHeartbeats(streamName);
+        }
+      } else {
+        long newSize = currentSize + dataSize;
+        success = absoluteSizes.replace(streamName, currentSize, newSize);
+      }
+    } while (!success);
+  }
+
+  protected HeartbeatPublisher getHeartbeatPublisher() {
+    return heartbeatPublisher;
+  }
+
+  protected int getInstanceId() {
+    return instanceId;
+  }
+
+  protected ConcurrentMap<String, Long> getAbsoluteSizes() {
+    return absoluteSizes;
+  }
+
+  protected ListeningScheduledExecutorService getScheduledExecutor() {
+    return scheduledExecutor;
+  }
+
+  /**
+   * Schedule publishing heartbeats for the {@code streamName}. At fixed rate, a heartbeat will be sent
+   * containing the absolute size of the files own by this stream handler and concerning the stream {@code streamName}.
+   */
+  protected void scheduleHeartbeats(final String streamName) {
+    scheduledExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
```

It would be a good idea to catch all exceptions here and log them, otherwise the Runnable will not be scheduled after an exception is thrown.